processor/processor.go

```diff
@@ -1894,41 +1902,50 @@ func (proc *Handle) Store(partition string, in *storeMessage) {
 	}
 
 	if len(destJobs) > 0 {
-		err := misc.RetryWithNotify(
-			context.Background(),
-			proc.jobsDBCommandTimeout,
-			proc.jobdDBMaxRetries,
-			func(ctx context.Context) error {
-				return proc.routerDB.WithStoreSafeTx(
-					ctx,
-					func(tx jobsdb.StoreSafeTx) error {
-						err := proc.routerDB.StoreInTx(ctx, tx, destJobs)
-						if err != nil {
-							return fmt.Errorf("storing router jobs: %w", err)
-						}
-
-						// rsources stats
-						err = proc.updateRudderSourcesStats(ctx, tx, destJobs)
-						if err != nil {
-							return fmt.Errorf("publishing rsources stats for router: %w", err)
-						}
-						return nil
-					})
-			}, proc.sendRetryStoreStats)
-		if err != nil {
-			panic(err)
-		}
-		proc.logger.Debug("[Processor] Total jobs written to router : ", len(destJobs))
+		func() {
+			// Only one goroutine can store to a router destination at a time, otherwise we may have different transactions
+			// committing at different timestamps which can cause events with lower jobIDs to appear after events with higher ones.
+			// For that purpose, before storing, we lock the relevant destination IDs (in sorted order to avoid deadlocks).
+			if len(in.routerDestIDs) > 0 {
+				destIDs := lo.Uniq(in.routerDestIDs)
+				slices.Sort(destIDs)
+				for _, destID := range destIDs {
+					proc.storePlocker.Lock(destID)
+					defer proc.storePlocker.Unlock(destID)
+				}
+			}
```

Comment:
Would it be simpler if we had just one lock? 

I haven't any benchmarks, but concurrent writes on the same table are not beneficial from a performance standpoint.

Reply:
We have observed major improvements in wait times when we have increased the number of writers, for some scenarios